Add K and Mg components to ASM2D

luohezhiming opened the pull request:
## Fixes/Resolves:

(replace this with the issue # fixed or resolved, if no issue exists then a brief statement of what this PR does)

## Summary/Motivation:
Add K and Mg as two inert components to ASM2D properties model, these two components are required to be tracked by ADM1 model of anaerobic digestion process, so we need to add these two components to complete the adm1/asm2d translator block #935 

## Changes proposed in this PR:
- add S_K and S_Mg to properties model
- update stoichiometry in reaction package 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

luohezhiming committed:
add

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
Add K and Mg components to ASM2D

luohezhiming committed:
delete redundant files

luohezhiming committed:
delete redundant test files

luohezhiming committed:
Merge branch 'main' into modified_asm2d

luohezhiming committed:
update flowsheets for ASM2d

luohezhiming committed:
revise flowsheet

luohezhiming committed:
revise inputs for S_K and S_Mg

luohezhiming committed:
add modified asm2d

luohezhiming committed:
Revise to directly inherent asm2d

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
add configuration option

luohezhiming committed:
Merge branch 'main' into modified_asm2d

MarcusHolly commented:
Pending any test failures, this looks really good to me. Although we should probably add documentation to this PR. It shouldn't require much work - essentially just copying and pasting the ASM2d documentation and then adding the color coding format that I've added to the Modified ADM1 documentation. Is this something you should handle @luohezhiming or should I just handle it?

adam-a-a commented:
> Pending any test failures, this looks really good to me. Although we should probably add documentation to this PR. It shouldn't require much work - essentially just copying and pasting the ASM2d documentation and then adding the color coding format that I've added to the Modified ADM1 documentation. Is this something you should handle @luohezhiming or should I just handle it?

I think that we want to let things play out before we commit to documenting this. For example, in this PR, we add K and Mg, but I don't think we added any associated constraints that tie to these newly added components. Before we dive into documentation, we should test what this will look like when we go to add constraints related to added components.

adam-a-a commented:
> > Pending any test failures, this looks really good to me. Although we should probably add documentation to this PR. It shouldn't require much work - essentially just copying and pasting the ASM2d documentation and then adding the color coding format that I've added to the Modified ADM1 documentation. Is this something you should handle @luohezhiming or should I just handle it?
> 
> I think that we want to let things play out before we commit to documenting this. For example, in this PR, we add K and Mg, but I don't think we added any associated constraints that tie to these newly added components. Before we dive into documentation, we should test what this will look like when we go to add constraints related to added components.

In other words, we may find out through experimentation that we don't want to go through with the options we chose.

luohezhiming commented:
> > Pending any test failures, this looks really good to me. Although we should probably add documentation to this PR. It shouldn't require much work - essentially just copying and pasting the ASM2d documentation and then adding the color coding format that I've added to the Modified ADM1 documentation. Is this something you should handle @luohezhiming or should I just handle it?
> 
> I think that we want to let things play out before we commit to documenting this. For example, in this PR, we add K and Mg, but I don't think we added any associated constraints that tie to these newly added components. Before we dive into documentation, we should test what this will look like when we go to add constraints related to added components.

I agree, we just add a configue to include more components, but no constraints add for this modified asm2d. We can dive into the doc later but if it get merged, we could finish translator blocks first

MarcusHolly commented:
What constraints need to be tied back to K and Mg? Is it essentially just the same constraints we added into Modified ADM1?
![image](https://user-images.githubusercontent.com/96305519/233629310-42d904ce-0f48-447b-979e-7f2758775e58.png)

luohezhiming committed:
Merge branch 'main' of https://github.com/watertap-org/watertap

luohezhiming committed:
Merge branch 'main' into modified_asm2d

luohezhiming commented:
> What constraints need to be tied back to K and Mg? Is it essentially just the same constraints we added into Modified ADM1? ![image](https://user-images.githubusercontent.com/96305519/233629310-42d904ce-0f48-447b-979e-7f2758775e58.png)

We don't have these constraints for ASM2D model

MarcusHolly commented:
> > What constraints need to be tied back to K and Mg? Is it essentially just the same constraints we added into Modified ADM1? ![image](https://user-images.githubusercontent.com/96305519/233629310-42d904ce-0f48-447b-979e-7f2758775e58.png)
> 
> We don't have these constraints for ASM2D model

I know, so my question is do we need to add these constraints. And do we need to add/modify any additional constraints?

luohezhiming commented:
> > > What constraints need to be tied back to K and Mg? Is it essentially just the same constraints we added into Modified ADM1? ![image](https://user-images.githubusercontent.com/96305519/233629310-42d904ce-0f48-447b-979e-7f2758775e58.png)
> > 
> > 
> > We don't have these constraints for ASM2D model
> 
> I know, so my question is do we need to add these constraints. And do we need to add/modify any additional constraints?

The reason to add this mol concentraiton for ADM1 is for H+ rule, I think if we don't have this rule for ASM2d, then no need to add these constraints

MarcusHolly commented on watertap/property_models/activated_sludge/modified_asm2d_properties.py:
```diff
@@ -0,0 +1,85 @@
+            self.add_component(j, Solute())
+
+        # self.S_K = Solute(doc="Potassium")
+        # self.S_Mg = Solute(doc="Magnesium")
```
No need for this to be here?

luohezhiming replied:
Good catch, I've deleted this.

luohezhiming committed:
delete redundant comments

adam-a-a commented:
Responding to @MarcusHolly 's question, my thinking was that as we start to add more of these components, there should be some incorporation those components into constraints. Although the base model for ASM2d may not have the same constraints that ADM1 has (e.g., charge balance), it will begin to need such additions as we add new components. THe point I wanted to make is that as we start to add new components on a conditional basis, we will also need to add associated constraints on a conditional basis (e.g., dissociation equations), which I suspect will get rather complicated.

adam-a-a commented on watertap/property_models/activated_sludge/modified_asm2d_properties.py:
```diff
@@ -0,0 +1,82 @@
+# "https://github.com/watertap-org/watertap/"
+#################################################################################
+"""
+Thermophysical property package to be used in conjunction with ASM2d reactions.
```
```suggestion
Thermophysical property package to be used in conjunction with modified ASM2d reactions for compatibility with the modified ADM1 model.
```

adam-a-a commented on watertap/property_models/activated_sludge/modified_asm2d_reactions.py:
```diff
@@ -0,0 +1,102 @@
+# "https://github.com/watertap-org/watertap/"
+#################################################################################
+"""
+ASM2d reaction package.
```
```suggestion
Modified ASM2d reaction package for compatibility with modified ADM1.
```

adam-a-a commented on watertap/property_models/activated_sludge/modified_asm2d_reactions.py:
```diff
@@ -0,0 +1,102 @@
+Important Note: ASM2d reactions depend on the presences of TSS in solution, however
+TSS does not take part in the rate expressions. Thus, it is possible to have cases
+where there is insufficient TSS present in the system for the reactions to occur
+resulting in an infeasible solution.
```
```suggestion
```

adam-a-a commented on watertap/property_models/activated_sludge/modified_asm2d_reactions.py:
```diff
@@ -0,0 +1,102 @@
+
+Reference:
+
+[1] Henze, M., Gujer, W., Mino, T., Matsuo, T., Wentzel, M.C., Marais, G.v.R.,
+Van Loosdrecht, M.C.M., "Activated Sludge Model No.2D, ASM2D", 1999,
+Wat. Sci. Tech. Vol. 39, No. 1, pp. 165-182
```
Should you add a reference showing the introduction of K and Mg into ASM2D?

luohezhiming replied:
The K and Mg are just added as two additional inert components and don't involve in any chemical precipitation process, so I don't have a good suggestion for reference of this modification.

MarcusHolly replied:
Yea, I agree that we don't need another reference here. Also, technically K and Mg aren't being added into this package at all - it's just the option to add any components that the user sees fit.

adam-a-a replied:
Right now you're just adding components as inert species, but later on, it is possible that we might have precipitation reactions and or a charge balance. My point being that we might as well add the Flores-Alsina refs that consider modification of ASM2d/ADM1 in the first place.

luohezhiming replied:
I added Flores_alsina ref

luohezhiming committed:
Update watertap/property_models/activated_sludge/modified_asm2d_properties.py

Co-authored-by: Adam Atia <aatia@keylogic.com>

luohezhiming committed:
Update watertap/property_models/activated_sludge/modified_asm2d_reactions.py

Co-authored-by: Adam Atia <aatia@keylogic.com>

luohezhiming committed:
Update watertap/property_models/activated_sludge/modified_asm2d_reactions.py

Co-authored-by: Adam Atia <aatia@keylogic.com>

MarcusHolly commented on watertap/property_models/activated_sludge/tests/test_modified_asm2d_reaction.py:
```diff
@@ -0,0 +1,932 @@
+        results = solver.solve(model, tee=True)
+        assert check_optimal_termination(results)
+
+    @pytest.mark.component
+    def test_solution(self, model):
```
Shouldn't we also test for the outlet concentration for Mg and K? It's currently missing from all the "test_solution"s

luohezhiming replied:
added

adam-a-a commented on watertap/property_models/activated_sludge/tests/test_modified_asm2d_reaction.py:
```diff
@@ -0,0 +1,932 @@
+# "https://github.com/watertap-org/watertap/"
+#################################################################################
+"""
+Tests for ASM2d reaction package.
```
```suggestion
Tests for modified ASM2d reaction package.
```

luohezhiming committed:
Update watertap/property_models/activated_sludge/tests/test_modified_asm2d_reaction.py

Co-authored-by: Adam Atia <aatia@keylogic.com>

luohezhiming committed:
add reference and test_solution

MarcusHolly reviewed (approved):
Everything LGTM now

adam-a-a reviewed (approved):
Approving this although I think we may need to revisit this approach once we move beyond getting the ASM2D/ADM1 translator to work.
